### ISSUE TITLE
Elastic search not working if sku is like 42-315

### ISSUE THREAD

**Issue by arijit-codilar** [magento/magento2#20689]
### Preconditions (*)

1. Magento 2.3 OpenSource
2. PHP 7.1
3. Apache
4. MySql 14.14
5. Ubuntu 18.04 LTS

### Steps to reproduce (*)

1. Create 2 or 3 products with sku like 42-315, 42-317 and 42-318.
2. Enable the elasticsearch in configuration and server.
3. Search in the frontend with sku 42-315.

### Expected result (*)

1. Only product with sku 42-315 should come.

### Actual result (*)

1. Unmatched products are also coming along with matched sku 42-315.
2. iPhone has sku 42-315.
![imagea](https://user-images.githubusercontent.com/46256180/51822428-4e2d2900-2302-11e9-8dc7-4abc8961b212.png)

**Comment by magento-engcom-team** [comment]
Hi @arijit-codilar. Thank you for your report. 
To help us process this issue please make sure that you provided the following information:
- [ ] Summary of the issue
- [ ] Information on your environment
- [ ] Steps to reproduce
- [ ] Expected and actual results

Please make sure that the issue is reproducible on the vanilla Magento instance following **Steps to reproduce**. To deploy vanilla Magento instance on our environment, please, add a comment to the issue: 

`@magento-engcom-team give me $VERSION instance` 

where `$VERSION` is [version tags](https://github.com/magento/magento2/tags) (starting from 2.2.0+) or develop branches (for example: 2.3-develop). 
For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation.

@arijit-codilar do you confirm that you was able to reproduce the issue on vanilla Magento instance following steps to reproduce?
- [ ] yes
- [ ] no

**Comment by magento-engcom-team** [comment]
Hi @LuciferStrome. Thank you for working on this issue. 
In order to make sure that issue has enough information and ready for development, please read and check the following instruction: :point_down:
- [ ] 1. Verify that issue has all the required information. (**Preconditions**, **Steps to reproduce**, **Expected result**, **Actual result**).<details><summary>Details</summary>If the issue has a valid description, the label `Issue: Format is valid` will be added to the issue automatically. Please, edit issue description if needed, until label `Issue: Format is valid` appears.</details>

- [ ] 2. Verify that issue has a meaningful description and provides enough information to reproduce the issue. If the report is valid, add `Issue: Clear Description` label to the issue by yourself.

- [ ] 3. Add `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.
- [ ] 4. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [ ] 5. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [ ] 6. Add label `Issue: Confirmed` once verification is complete. 

- [ ] 7. Make sure that automatic system confirms that report has been added to the backlog.

**Comment by magento-engcom-team** [comment]
Hi @engcom-backlog-nazar. Thank you for working on this issue. 
In order to make sure that issue has enough information and ready for development, please read and check the following instruction: :point_down:
- [ ] 1. Verify that issue has all the required information. (**Preconditions**, **Steps to reproduce**, **Expected result**, **Actual result**).<details><summary>Details</summary>If the issue has a valid description, the label `Issue: Format is valid` will be added to the issue automatically. Please, edit issue description if needed, until label `Issue: Format is valid` appears.</details>

- [ ] 2. Verify that issue has a meaningful description and provides enough information to reproduce the issue. If the report is valid, add `Issue: Clear Description` label to the issue by yourself.

- [ ] 3. Add `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.
- [ ] 4. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [ ] 5. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [ ] 6. Add label `Issue: Confirmed` once verification is complete. 

- [ ] 7. Make sure that automatic system confirms that report has been added to the backlog.

**Comment by ghost** [comment]
HI @ arijit-codilar thank you for your report, just for note, if you use Mysql search engine and type on search "42-" the exception is become.
![selection_295](https://user-images.githubusercontent.com/40756439/51825325-2f706780-22ed-11e9-874d-fbce5f44009b.png)

**Comment by arijit-codilar** [comment]
@engcom-backlog-nazar I am getting the below result when I am searching with 42- using MySql which is again I believe an issue because it should return the products matching that query text.

![imageb](https://user-images.githubusercontent.com/46256180/51825943-e4b21800-230b-11e9-8b0f-59db9c56a027.png)

**Comment by ghost** [comment]
@arijit-codilar hmm.. strange i'm getting exception on clean 2.3-dev instance

**Comment by arijit-codilar** [comment]
@engcom-backlog-nazar I have installed the Magento from magento's site but still it should not get into two directions.

**Comment by magento-engcom-team** [comment]
@engcom-backlog-nazar Thank you for verifying the issue. Based on the provided information internal tickets `MAGETWO-97943`, `MAGETWO-97944` were created

**Comment by ghost** [comment]
@magento-engcom-team give me 2.3-develop  instance

**Comment by magento-engcom-team** [comment]
Hi @engcom-backlog-nazar. Thank you for your request. I'm working on Magento 2.3-develop  instance for you

**Comment by amol2jcommerce** [comment]
@magento-engcom-team give me 2.3-develop instance

**Comment by magento-engcom-team** [comment]
Hi @amol2jcommerce. Thank you for your request. I'm working on Magento 2.3-develop instance for you

**Comment by magento-engcom-team** [comment]
Hi @amol2jcommerce, here is your [Magento instance](https://i-20689-2-3-develop.instances.magento-community.engineering).
Admin access: [https://i-20689-2-3-develop.instances.magento-community.engineering/admin](https://i-20689-2-3-develop.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`
Instance will be terminated in up to 3 hours.

**Comment by ghost** [comment]
@arijit-codilar on vanilla 
![selection_296](https://user-images.githubusercontent.com/40756439/51826625-2cc34180-22f0-11e9-81f6-14fb63a0640c.png)

**Comment by arijit-codilar** [comment]
@engcom-backlog-nazar But I am getting something different which I have shared above. And, my instance has been downloaded from magento's site. 

Are you using Elastic search or MySql?

**Comment by ghost** [comment]
@arijit-codilar no this only appears on Mysql, but perfectly works with Elastic

**Comment by arijit-codilar** [comment]
@engcom-backlog-nazar  My issue is with elastic search. I have two products iPhone 6S -> SKU - 42-315 and iPhone 7S -> SKU - 42-317. Now, if I search with 42-315, it should return iPhone 6S not other products but it is returning other unmatched products too. 

See the description for more clarity.

**Comment by ghost** [comment]
@arijit-codilar yes, yours issue confirmed, i'm trying to say that there is another issue with related with yours.

**Comment by arijit-codilar** [comment]
@engcom-backlog-nazar Ok. With the mysql search engine, the exception that you are getting for that are the products with sku 42-315 like that created in admin? or you are searching without creating the products in backend?

**Comment by ghost** [comment]
@arijit-codilar The products are created in admin, you can check on this link https://i-20689-2-3-develop.instances.magento-community.engineering/ just type search like '42-'

**Comment by arijit-codilar** [comment]
@engcom-backlog-nazar Ya. I can see the product in admin and the issue also I can see now. So, yeah, its another issue.

**Comment by arijit-codilar** [comment]
@engcom-backlog-nazar If I search with 42-315, then also no product is showing up. I guess its due to that the product is not assigned to any category? Any say?

**Comment by Nazar65** [comment]
@arijit-codilar i'm found solution for elastic, you need to edit mysql conf and set this variables 
```
[mysqld]
innodb_ft_min_token_size=2
ft_min_word_len=2
```
the you have search result and first item is ehat you requested.

**Comment by arijit-codilar** [comment]
@Nazar65 its not working with the mysql.conf changes you have shared.

**Comment by magento-engcom-team** [comment]
Hi @arijit-codilar. Thank you for your report. 
The issue has been fixed in magento/magento2#20727 by @Nazar65 in 2.3-develop branch
Related commit(s):
 - [cb681ab9f7f44457b420d7ae86cf7b324b7b8c94](https://github.com/magento/magento2/commit/cb681ab9f7f44457b420d7ae86cf7b324b7b8c94)
 - [7539e8e6829197467b59e417e317d38ea39a005b](https://github.com/magento/magento2/commit/7539e8e6829197467b59e417e317d38ea39a005b)

The fix will be available with the upcoming 2.3.1 release.

**Comment by arijit-codilar** [comment]
@magento-engcom-team The issue is still not resolved. The changes that he made is already there in Magento 2.3. Please reopen the issue.

**Comment by arijit-codilar** [comment]
@Nazar65 The changes that you made are already there in Magento 2.3. With this elastic search is still not working. Please download Magento 2.3 from magento.com and verify once.
See the screenshot attached.
SynonymReader
![synonymreader](https://user-images.githubusercontent.com/46256180/52548682-18556d80-2df5-11e9-8256-812329d04810.png)
SynonymAnalyzer
![synonymanalyzer](https://user-images.githubusercontent.com/46256180/52548689-1f7c7b80-2df5-11e9-916a-a41d2d90ff11.png)
Match
![match](https://user-images.githubusercontent.com/46256180/52548693-23100280-2df5-11e9-9114-8dff3a39fbb5.png)

**Comment by magento-engcom-team** [comment]
Hi @Nazar65. Thank you for working on this issue. 
Looks like this issue is already verified and confirmed. But if you want to validate it one more time, please, go though the following instruction:
- [ ] 1. Add/Edit `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.

- [ ] 2. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [ ] 3. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [ ] 4. If the issue is not relevant or is not reproducible any more, feel free to close it.

**Comment by Nazar65** [comment]
@arijit-codilar this wrongly added to the referenced pr but i'm working on this with elastic search

**Comment by lfluvisotto** [comment]
@magento-engcom-team give me 2.3.1

**Comment by lfluvisotto** [comment]
@magento-engcom-team give me 2.3.1 instance

**Comment by magento-engcom-team** [comment]
Hi @lfluvisotto. Thank you for your request. I'm working on Magento 2.3.1 instance for you

**Comment by magento-engcom-team** [comment]
Hi @lfluvisotto, here is your [Magento instance](https://i-20689-2-3-1.instances.magento-community.engineering).
Admin access: [https://i-20689-2-3-1.instances.magento-community.engineering/admin](https://i-20689-2-3-1.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`
Instance will be terminated in up to 3 hours.

**Comment by geet07** [comment]
@magento-engcom-team give me 2.3.1 instance

**Comment by magento-engcom-team** [comment]
Hi @geet07. Thank you for your request. I'm working on Magento 2.3.1 instance for you

**Comment by magento-engcom-team** [comment]
Hi @geet07, here is your [Magento instance](https://i-20689-2-3-1.instances.magento-community.engineering).
Admin access: [https://i-20689-2-3-1.instances.magento-community.engineering/admin](https://i-20689-2-3-1.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`
Instance will be terminated in up to 3 hours.

**Comment by lfluvisotto** [comment]
@magento-engcom-team

It would be great if the instance comes with elasticseach setup by default

I'm unable to test this issue with vanilla instance

**Comment by Nazar65** [comment]
@lfluvisotto https://github.com/magento/community-features/issues/145

**Comment by lfluvisotto** [comment]
@Nazar65, nice! I was't knowing about that, thanks for the tip.

**Comment by denchev** [comment]
@magento give me 2.4-develop instance

**Comment by denchev** [comment]
This behavior continues to appear in 2.4 Not quite sure if this is a bug but I will spend some time to investigate a bit deeper. 

I created 3 products with the mentioned skus 42-315, 42-317 and 42-318. When I search for the first one 42-315 I got all the 3 products + 3 more than have SKUs like 24-WG080, 24-WG088 and 24-WG084. Since 24 is the reverse of 42 and there is a dash in the SKU I assume that Elasticsearch is doing some "magic" when storing the results. 

I will try to see if there are any settings that can be tweaked.

@magento I am working on this

**Comment by denchev** [comment]
I've been investigating this and it looks like the problem is with [Elasticsearch tokenizer](https://www.elastic.co/guide/en/elasticsearch/reference/current/analysis-standard-tokenizer.html). I would not count this is a bug per-say as this is how Elasticsearch is supposed to work. I found this to be a merchant-facing issue as I had clients that complain about this same behavior. I am working on a potential solution - a non-core module (for now) that will allow merchants to turn on/off tokenization for certain attributes (sku). This should fix this type of issue. I don't think this is something that belongs to Magento 2 core code. As soon as I have it I will post it here and maybe then a discussion can start on what to do with the whole problem in general.

**Comment by magento-engcom-team** [comment]
Hi @arijit-codilar, @denchev. 

Thank you for your report and collaboration!

The related internal Jira ticket `MC-24438` was closed as `Fixed`. 

The fix will be available with the upcoming `2.4.3` release.
